File: python/visio_schema/transport/framing.py
"""COBS-delimited core-frame de/framing — the single Python implementation.

Serial/byte transports carry COBS-delimited core frames (framing.md §3.2). This
is the one place the split-on-0x00 + COBS-decode + frame-decode loop lives;
:class:`SerialEndpoint`, the byte-stream reader (:func:`read_frames`), and the
examples all use it instead of hand-rolling it.
"""
from __future__ import annotations

import logging
from collections.abc import Iterable, Iterator

from visio_schema.wire.codec import cobs_decode, cobs_encode, decode_frame, encode_frame
from visio_schema.wire.message import Message

log = logging.getLogger(__name__)


def frame_bytes(msg: Message) -> bytes:
    """Serialize a Message to its on-the-wire COBS frame (with the 0x00
    delimiter): ``COBS(HEADER_LEN || header || payload || CRC) || 0x00``."""
    return cobs_encode(encode_frame(msg.to_header(), msg.payload)) + b"\x00"


def extract_frames(rx_buf: bytearray) -> list[Message]:
    """Pull every complete 0x00-delimited frame out of ``rx_buf`` (consuming the
    bytes it processes, leaving any partial trailing frame) and return the
    decoded Messages. Malformed frames are logged and skipped (framing.md §5)."""
    msgs: list[Message] = []
    # Advance a cursor and delete consumed bytes ONCE at the end, rather than an
    # O(remaining) ``del rx_buf[:n]`` shift per frame.
    pos = 0
    while True:
        delim = rx_buf.find(b"\x00", pos)
        if delim < 0:
            break
        encoded = bytes(rx_buf[pos:delim])
        pos = delim + 1
        if not encoded:
            continue
        try:
            header, payload = decode_frame(cobs_decode(encoded))
        except Exception as exc:
            log.warning("dropping malformed frame: %s", exc)
            continue
        msgs.append(Message.from_header(header, payload))
    if pos:
        del rx_buf[:pos]
    return msgs


def read_frames(chunks: Iterable[bytes]) -> Iterator[Message]:
    """Yield Messages from an iterable of byte chunks (e.g. successive serial
    reads), buffering partial frames across chunk boundaries. The byte-stream
    convenience over :func:`extract_frames`; transport (opening the port) stays
    with the caller."""
    buf = bytearray()
    for chunk in chunks:
        if chunk:
            buf.extend(chunk)
        yield from extract_frames(buf)

File: cpp/src/transport/framing.cc
#include "visio_schema/transport/framing.hpp"

#include <algorithm>
#include <iostream>
#include <string>
#include <string_view>

#include "visio_schema/wire/codec/cobs.hpp"
#include "visio_schema/wire/codec/frame.hpp"

namespace visio_schema::transport {

std::vector<Message> ExtractFrames(std::vector<std::uint8_t>& rx_buf) {
  std::vector<Message> out;
  while (true) {
    auto it = std::find(rx_buf.begin(), rx_buf.end(), std::uint8_t{0});
    if (it == rx_buf.end()) break;
    std::size_t delim = static_cast<std::size_t>(it - rx_buf.begin());
    if (delim == 0) {
      rx_buf.erase(rx_buf.begin());  // bare 0x00; skip empty frame
      continue;
    }
    std::vector<std::uint8_t> encoded(rx_buf.begin(), rx_buf.begin() + delim);
    rx_buf.erase(rx_buf.begin(), rx_buf.begin() + delim + 1);
    std::vector<std::uint8_t> decoded;
    decoded.reserve(encoded.size());
    std::string_view enc_view{reinterpret_cast<const char*>(encoded.data()),
                              encoded.size()};
    if (!visio_schema::wire::CobsDecode(enc_view, &decoded)) {
      std::cerr << "visio-schema: COBS decode failed (" << encoded.size()
                << " bytes)\n";
      continue;
    }
    Message msg;
    std::string_view frame_view{reinterpret_cast<const char*>(decoded.data()),
                                decoded.size()};
    const auto status = visio_schema::wire::DecodeFrame(frame_view, &msg);
    if (status != visio_schema::wire::FrameStatus::kOk) {
      std::cerr << "visio-schema: frame decode dropped: "
                << visio_schema::wire::FrameStatusName(status) << "\n";
      continue;
    }
    out.push_back(std::move(msg));
  }
  return out;
}

bool WriteFramed(Link& link, const Message& msg) {
  const std::string frame = visio_schema::wire::EncodeFrame(msg);
  auto encoded = visio_schema::wire::CobsEncode(std::string_view{frame});
  encoded.push_back(0);
  return link.Write(
      {reinterpret_cast<const char*>(encoded.data()), encoded.size()});
}

}  // namespace visio_schema::transport
